block/submitter.go

```diff
@@ -18,88 +17,118 @@ func (m *Manager) HeaderSubmissionLoop(ctx context.Context) {
 	for {
 		select {
 		case <-ctx.Done():
+			m.logger.Info("header submission loop stopped")
 			return
 		case <-timer.C:
 		}
 		if m.pendingHeaders.isEmpty() {
 			continue
 		}
-		err := m.submitHeadersToDA(ctx)
+		headersToSubmit, err := m.pendingHeaders.getPendingHeaders(ctx)
+		if err != nil {
+			m.logger.Error("error while fetching headers pending DA", "err", err)
+			continue
+		}
+		if len(headersToSubmit) == 0 {
+			continue
+		}
+		err = m.submitHeadersToDA(ctx, headersToSubmit)
 		if err != nil {
 			m.logger.Error("error while submitting header to DA", "error", err)
 		}
 	}
 }
 
-func (m *Manager) submitHeadersToDA(ctx context.Context) error {
-	submittedAllHeaders := false
-	var backoff time.Duration
-	headersToSubmit, err := m.pendingHeaders.getPendingHeaders(ctx)
-	if len(headersToSubmit) == 0 {
-		// There are no pending headers; return because there's nothing to do, but:
-		// - it might be caused by error, then err != nil
-		// - all pending headers are processed, then err == nil
-		// whatever the reason, error information is propagated correctly to the caller
-		return err
-	}
-	if err != nil {
-		// There are some pending headers but also an error. It's very unlikely case - probably some error while reading
-		// headers from the store.
-		// The error is logged and normal processing of pending headers continues.
-		m.logger.Error("error while fetching headers pending DA", "err", err)
+// DataSubmissionLoop is responsible for submitting data to the DA layer.
+func (m *Manager) DataSubmissionLoop(ctx context.Context) {
+	timer := time.NewTicker(m.config.DA.BlockTime.Duration)
+	defer timer.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			m.logger.Info("data submission loop stopped")
+			return
+		case <-timer.C:
+		}
+		if m.pendingData.isEmpty() {
+			continue
+		}
+
+		signedDataToSubmit, err := m.createSignedDataToSubmit(ctx)
+		if err != nil {
+			m.logger.Error("failed to create signed data to submit", "error", err)
+			continue
+		}
+		if len(signedDataToSubmit) == 0 {
+			continue
+		}
+
+		err = m.submitDataToDA(ctx, signedDataToSubmit)
+		if err != nil {
+			m.logger.Error("failed to submit data to DA", "error", err)
+		}
 	}
-	numSubmittedHeaders := 0
+}
+
+// submitToDA is a generic helper for submitting items to the DA layer with retry, backoff, and gas price logic.
+// marshalFn marshals an item to []byte.
+// postSubmit is called after a successful submission to update caches, pending lists, etc.
+func submitToDA[T any](
+	m *Manager,
+	ctx context.Context,
+	items []T,
+	marshalFn func(T) ([]byte, error),
```

Comment:
it doesnt seem like we do anything specific with the items so can we provide it to the function as encoded bytes already?

Reply:
Explored this a bit, looks like the `postSubmit` method expects the items in not encoded form, so it's not as straightforward as passing the bytes encoded. Created follow up issue here: #2385